File: 2018/day06/src/main.rs
use std::collections::{HashMap, HashSet};
use std::fs;

type Point = (usize, usize);
type PointMap = HashMap<Point, usize>;
type Grid = Vec<Vec<(usize, bool)>>;

fn main() {
    let input = fs::read_to_string("input").expect("Couldn't read file");

    let mut points = parse_input(&input);

    let max_x = points.keys().map(|p| p.0).max().unwrap();
    let max_y = points.keys().map(|p| p.1).max().unwrap();
    let mut grid: Grid = vec![vec![(0, false); max_y + 2]; max_x + 2];

    while points.len() > 0 {
        for (p, k) in points.iter() {
            grid[p.0][p.1] = (*k, true);
        }

        points = next_points(points, &grid, max_x, max_y);
    }

    let counts = get_counts(&grid);
    let max_area = counts.iter().map(|(_, v)| v).max().unwrap();

    println!("Part one: {}", max_area);

    let points: HashSet<Point> = parse_input(&input).keys().map(|&k| k).collect();
    println!("Part two: {}", safe_points(points, max_x, max_y));

    // let mut grid: [[(usize, bool); 320]; 320] = [[(0, false); MAX_X]; MAX_Y];
    // let point = parse_input(&input);

    // println!("Used to calculate grid size");
    // println!("{} Starting points", input.len());
    // println!("Min x: {:?}", input.iter().min_by_key(|x| x.1).unwrap().1);
    // println!("Max x: {:?}", input.iter().max_by_key(|x| x.1).unwrap().1);
    // println!("Min y: {:?}", input.iter().min_by_key(|x| x.2).unwrap().2);
    // println!("Max y: {:?}", input.iter().max_by_key(|x| x.2).unwrap().2);
}

fn safe_points(points: HashSet<Point>, max_x: usize, max_y: usize) -> usize {
    let mut count = 0;

    for x in 0..max_x {
        for y in 0..max_y {
            let mut sum = 0;
            for p in points.iter() {
                sum += if p.0 > x { p.0 - x } else { x - p.0 };
                sum += if p.1 > y { p.1 - y } else { y - p.1 };
            }
            if sum < 10000 {
                count += 1;
            }
        }
    }

    count
}

fn get_counts(grid: &Grid) -> HashMap<usize, usize> {
    let edges = get_edge_pieces(grid);

    let mut counts: HashMap<usize, usize> = HashMap::new();
    for row in grid.iter() {
        for &(v, _) in row.iter() {
            if !edges.contains(&v) {
                counts.insert(v, counts.get(&v).unwrap_or(&0) + 1);
            }
        }
    }
    counts
}

fn get_edge_pieces(grid: &Grid) -> HashSet<usize> {
    let mut edges: HashSet<usize> = HashSet::new();

    for k in grid[0].iter() {
        edges.insert(k.0);
    }

    for row in grid.iter() {
        let mut row = row.iter();
        edges.insert(row.next().unwrap().0);
        edges.insert(row.last().unwrap().0);
    }

    for k in grid[grid.len() - 1].iter() {
        edges.insert(k.0);
    }

    edges
}

fn parse_input(input: &str) -> PointMap {
    let points: Vec<Point> = input
        .lines()
        .map(|s| {
            let coords: Vec<usize> = s.split(", ").map(|c| c.parse().unwrap()).collect();
            (coords[0], coords[1])
        })
        .collect();

    let min_x = points.iter().map(|p| p.0).min().unwrap();
    let min_y = points.iter().map(|p| p.1).min().unwrap();

    points
        .into_iter()
        .enumerate()
        .map(|(i, p)| ((p.0 + 1 - min_x, p.1 + 1 - min_y), i + 1))
        .collect()
}

fn next_points(points: PointMap, grid: &Grid, max_x: usize, max_y: usize) -> PointMap {
    let mut next: PointMap = HashMap::new();

    for (p, k) in points {
        if k == 0 {
            continue;
        }

        for n in neighbours(&p, max_x, max_y) {
            let assigned = grid[n.0][n.1].1;
            if assigned {
                continue;
            }

            if next.get(&n).is_some() && *next.get(&n).unwrap() != k {
                next.insert(n, 0);
                continue;
            }

            next.insert(n, k);
        }
    }

    next
}

fn neighbours(p: &Point, max_x: usize, max_y: usize) -> Vec<Point> {
    let mut n = Vec::new();
    if p.0 > 0 {
        n.push((p.0 - 1, p.1));
    }
    if p.0 <= max_x {
        n.push((p.0 + 1, p.1));
    }
    if p.1 > 0 {
        n.push((p.0, p.1 - 1));
    }
    if p.1 <= max_y {
        n.push((p.0, p.1 + 1));
    }

    n
}

File: 2018/day06/main.py
from collections import defaultdict


def read_input():

    points = []
    with open("test_input") as f:
        for line in f:
            x, y = [int(x) for x in line.split(", ")]
            points.append((x, y))

    # Normalise so mininum value of x and y is 1
    min_x = min([p[0] for p in points])
    min_y = min([p[1] for p in points])

    return {(p[0] - min_x + 1, p[1] - min_y + 1): i+1 for (i, p) in enumerate(points)}


def make_grid(points):

    max_x = max([p[0] for p in points.keys()])
    max_y = max([p[1] for p in points.keys()])
    print("Grid size:", max_x, "x", max_y)

    grid = [[(0, False) for i in range(max_y + 2)] for y in range(max_x + 2)]

    while points:
        current = {}

        # Set grid for previous iterations's points
        for ((x, y), i) in points.items():
            grid[x][y] = (i, True)

        for ((x, y), i) in points.items():
            if i == 0:
                continue

            for (m, n) in [(x+1, y), (x-1, y), (x, y+1), (x, y-1)]:

                # If on an edge continue
                if m < 0 or m > max_x + 1 or n < 0 or n > max_y + 1:
                    continue

                # If already assigned, skip
                try:
                    if grid[m][n][1]:
                        continue
                except:
                    print(m, n)

                if (m, n) in current and current[(m, n)] != i:
                    current[(m, n)] = 0
                else:
                    current[(m, n)] = i

        points = current

    return grid


def sum_areas(grid):
    areas = defaultdict(int)

    for row in grid:
        for (p, _) in row:
            areas[p] += 1

    return areas


def edge_pieces(grid):

    edges = set([p[0] for p in grid[0]] + [p[0] for p in grid[-1]])
    for row in grid:
        edges.add(row[0][0])
        edges.add(row[-1][0])

    return edges


def solve_1():

    points = read_input()
    grid = make_grid(points)
    print(grid)

    areas = sum_areas(grid)
    print(areas)

    edges = edge_pieces(grid)
    print(edges)

    answer = max([v for (k, v) in areas.items() if k not in edges])
    print(answer)


def solve_2():

    # points = {
    #     (1, 1): 'a',
    #     (1, 6): 'b',
    #     (8, 3): 'c',
    #     (3, 4): 'd',
    #     (5, 5): 'e',
    #     (8, 9): 'f',
    # }
    points = read_input()

    count = 0

    max_x = max([p[0] for p in points.keys()])
    max_y = max([p[1] for p in points.keys()])

    for x in range(max_x):
        for y in range(max_y):

            total = 0
            for (m, n) in points.keys():
                total += abs(x - m) + abs(y - n)

            if total < 10000:
                count += 1

    print(count)


if __name__ == "__main__":
    solve_1()
    solve_2()
